***Bishal Kundu***

***Encryptix | Data Science Internship***

***Titanic Survival Prediction | | Task 1***

In [1]:
#Importing Libraries

import pandas as pd
from sklearn.linear_model import LogisticRegression as Logreg
from sklearn.preprocessing import LabelEncoder as Le
from sklearn.model_selection import train_test_split as tt
from sklearn.metrics import accuracy_score, confusion_matrix

#Importing Titanic survival dataset

data=pd.read_csv('/content/Titanic-Dataset.csv')
data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [2]:
#Data Count

print(data.count())

# Duplicate value finding from the primary key

print('Duplicate=',data['PassengerId'].duplicated().sum())

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64
Duplicate= 0


**There was no duplicate IDs in the primary key so we proceed with the analysis**

In [3]:
#Null Value Finding

data.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

In [4]:
#Replacing null ages with mean value

data['Age']=data['Age'].fillna(data['Age'].mean())

#New available number of Null vaules

data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

**The null values of age is replaced b the mean age to make the prediction better**

In [5]:
#Count of Embarked category

data['Embarked'].value_counts()

Embarked
S    644
C    168
Q     77
Name: count, dtype: int64

In [6]:
#The mode value for Embarked is 'S'(644), Replacing Null with mode

data['Embarked']=data['Embarked'].fillna('S')

#New Null value

data.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age              0
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         0
dtype: int64

**The null values in Embarked variable was replaced with the mode or most repetative port**

In [7]:
#Transforming string to numeric value with LabelEncoder

data['Embarked_Num']=Le().fit_transform(data['Embarked'])
data['Sex_Num']=Le().fit_transform(data['Sex'])
data

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex        Age  \
0                              Braund, Mr. Owen Harris    male  22.000000   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.000000   
2                               Heikkinen, Miss. Laina  female  26.000000   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.000000   
4                             Allen, Mr. William Henry    male  35.000000   
..                                                 ...     ...        ...   
886                              Montvila, Rev. Juozas    male  27.000000   
887                       Graham, Miss. Margaret Edith  female  19.000000   
888           Johnston, Miss. Catherine Helen "Carrie"  female  29.699118   
889                              Behr, Mr. Karl Howell    male  26.000000   
890                                Dooley, Mr. Patrick    male  32.000000   

     SibSp  Parch            Ticket     Fare Cabin Embarked  Embarked_Num  \
0        1      0         A/5 21171   7.2500   NaN        S             2   
1        1      0          PC 17599  71.2833   C85        C             0   
2        0      0  STON/O2. 3101282   7.9250   NaN        S             2   
3        1      0            113803  53.1000  C123        S             2   
4        0      0            373450   8.0500   NaN        S             2   
..     ...    ...               ...      ...   ...      ...           ...   
886      0      0            211536  13.0000   NaN        S             2   
887      0      0            112053  30.0000   B42        S             2   
888      1      2        W./C. 6607  23.4500   NaN        S             2   
889      0      0            111369  30.0000  C148        C             0   
890      0      0            370376   7.7500   NaN        Q             1   

     Sex_Num  
0          1  
1          0  
2          0  
3          0  
4          1  
..       ...  
886        1  
887        0  
888        0  
889        1  
890        1  

[891 rows x 14 columns]

**The data values for Sex and Embarked was updated to numbers to bring these variables under the predictive Model**

In [8]:
#Removing unnecessary columns

data.drop((['PassengerId','Name','Sex','Embarked','Ticket','Cabin','Fare']), axis=1,inplace=True)
data

Survived  Pclass        Age  SibSp  Parch  Embarked_Num  Sex_Num
0           0       3  22.000000      1      0             2        1
1           1       1  38.000000      1      0             0        0
2           1       3  26.000000      0      0             2        0
3           1       1  35.000000      1      0             2        0
4           0       3  35.000000      0      0             2        1
..        ...     ...        ...    ...    ...           ...      ...
886         0       2  27.000000      0      0             2        1
887         1       1  19.000000      0      0             2        0
888         0       3  29.699118      1      2             2        0
889         1       1  26.000000      0      0             0        1
890         0       3  32.000000      0      0             1        1

[891 rows x 7 columns]

**Only the necessery columns were kept for further analysis and for building the predictive model**

In [9]:
#Data Statistical Description

data.describe()

Survived      Pclass         Age       SibSp       Parch  \
count  891.000000  891.000000  891.000000  891.000000  891.000000   
mean     0.383838    2.308642   29.699118    0.523008    0.381594   
std      0.486592    0.836071   13.002015    1.102743    0.806057   
min      0.000000    1.000000    0.420000    0.000000    0.000000   
25%      0.000000    2.000000   22.000000    0.000000    0.000000   
50%      0.000000    3.000000   29.699118    0.000000    0.000000   
75%      1.000000    3.000000   35.000000    1.000000    0.000000   
max      1.000000    3.000000   80.000000    8.000000    6.000000   

       Embarked_Num     Sex_Num  
count    891.000000  891.000000  
mean       1.536476    0.647587  
std        0.791503    0.477990  
min        0.000000    0.000000  
25%        1.000000    0.000000  
50%        2.000000    1.000000  
75%        2.000000    1.000000  
max        2.000000    1.000000

**Age Statistics (Age):**

Mean: 29.699118 years

Min: 0.42 years

Max: 80.00 years


**Number of Siblings/Spouses Aboard (SibSp):**


Min: 0

Max: 8

**Number of Parents/Children Aboard (Parch):**

Min: 0

Max: 6

In [10]:
#Data Information

data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Survived      891 non-null    int64  
 1   Pclass        891 non-null    int64  
 2   Age           891 non-null    float64
 3   SibSp         891 non-null    int64  
 4   Parch         891 non-null    int64  
 5   Embarked_Num  891 non-null    int64  
 6   Sex_Num       891 non-null    int64  
dtypes: float64(1), int64(6)
memory usage: 48.9 KB


In [11]:
# Definig Dependent and Independent variables

x=data.drop(['Survived'],axis=1)
y=data['Survived']
x,y

(     Pclass        Age  SibSp  Parch  Embarked_Num  Sex_Num
 0         3  22.000000      1      0             2        1
 1         1  38.000000      1      0             0        0
 2         3  26.000000      0      0             2        0
 3         1  35.000000      1      0             2        0
 4         3  35.000000      0      0             2        1
 ..      ...        ...    ...    ...           ...      ...
 886       2  27.000000      0      0             2        1
 887       1  19.000000      0      0             2        0
 888       3  29.699118      1      2             2        0
 889       1  26.000000      0      0             0        1
 890       3  32.000000      0      0             1        1
 
 [891 rows x 6 columns],
 0      0
 1      1
 2      1
 3      1
 4      0
       ..
 886    0
 887    1
 888    0
 889    1
 890    0
 Name: Survived, Length: 891, dtype: int64)

In [12]:
#Training the Logistic model

x_train,x_test,y_train,y_test=tt(x,y,test_size=0.2,random_state=30)
model=Logreg()
model.fit(x_train,y_train)


LogisticRegression()

**The model was trained with 20% of the data**

In [13]:
#Predicting value with text data

y_pred=model.predict(x_test)
y_pred

array([1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 1,
       1, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0,
       0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1,
       0, 1, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 1,
       1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0,
       1, 1, 0])

In [14]:
#Checking Accuracy and Confusion matrix

print('Accuracy=',accuracy_score(y_test,y_pred))
print('Confusion Matrix=',confusion_matrix(y_test,y_pred))

Accuracy= 0.776536312849162
Confusion Matrix= [[92 16]
 [24 47]]


**The accuracy score between the test and prediction data is 77% and the confusion matrix suggests,**

Correctly identified 92 positive cases (True Positives).

Incorrectly identified 16 negative cases as positive (False Positives).

Incorrectly identified 24 positive cases as negative (False Negatives).

Correctly identified 47 negative cases (True Negatives).

In [15]:
#Predicing Survival for Passenger class for Male

print('Case 1=',model.predict([[1,40.0,1,5,2,1]]))
print('Case 2=',model.predict([[2,40.0,1,5,2,1]]))
print('Case 3=',model.predict([[3,40.0,1,5,2,1]]))

Case 1= [0]
Case 2= [0]
Case 3= [0]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


**Predicting the male survival data based on their class, he is 40 of age, no
sibling or spouse, one parent or child, from Southampton town.**

**For all the classes, male suvival rate was zero.**

In [16]:
#Predicing Survival for Passenger class for Female

print('Case 1=',model.predict([[1,40.0,1,5,2,0]]))
print('Case 2=',model.predict([[2,40.0,1,5,2,0]]))
print('Case 3=',model.predict([[3,40.0,1,5,2,0]]))

Case 1= [1]
Case 2= [1]
Case 3= [0]


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
/usr/local/lib/python3.10/dist-packages/sklearn/base.py:465: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


**Predicting the female survival data based on their class, he is 40 of age, no
sibling or spouse, one parent or child, from Southampton town.**

**For 1st and 2nd class, females suvive and 3rd class felmale die.**

**Thank You**